extensions/amp-story/0.1/amp-story-auto-ads.js

```diff
@@ -29,42 +34,151 @@ export class AmpStoryAutoAds extends AMP.BaseElement {
   constructor(element) {
     super(element);
 
-    /** @private {?AmpStory} */
+    /** @private {?./amp-story.AmpStory} */
     this.ampStory_ = null;
+
+    /** @private {?./navigation-state.NavigationState} */
+    this.navigationState_ = null;
+
+    /** @private {number} */
+    this.interactions_ = 0;
+
+    /** @private {!Array} */
+    this.adElements_ = [];
+
+    /** @private {number} */
+    this.adIdPointer_ = 0;
+
+    /** @private {?Array} */
+    this.pages_ = null;
   }
 
   /** @override */
   buildCallback() {
-    const ampStory = this.element.parentElement;
-    user().assert(ampStory.tagName === 'AMP-STORY',
+    const ampStoryElement = this.element.parentElement;
+    user().assert(ampStoryElement.tagName === 'AMP-STORY',
         `<${TAG}> should be child of <amp-story>`);
-    this.ampStory_ = ampStory;
-  }
+
+    ampStoryElement.getImpl().then(impl => {
+      this.ampStory_ = impl;
+      this.navigationState_ = this.ampStory_.getNavigationState();
+      this.navigationState_.observe(this.handleStateChange_.bind(this));
+    });
+
+    // TODO(ccordry): move this chunk to layoutCallback
+    // TODO(ccordry): need to fake distance from page to force load
+    for (let i = 0; i < 2; i++) {
+      const mockPage = this.makeMockPage();
+      this.adElements_.push(mockPage);
+      ampStoryElement.appendChild(mockPage);
+    }
+  };
+
 
   /** @override */
   isLayoutSupported() {
     return true;
   }
 
+
   /** @override */
-  layoutCallback() {
-    this.ampStory.appendChild(this.makeMockPage());
-  }
+  // layoutCallback() {}
+
 
   // temporary to be replaced with real page later
   makeMockPage() {
     const ampStoryAdPage = document.createElement('amp-story-page');
-    ampStoryAdPage.id = 'i-amphtml-ad-page-1';
+    const id = this.adElements_.length + 1;
+    ampStoryAdPage.id = `i-amphtml-ad-page-${id}`;
+    ampStoryAdPage.setAttribute('advertisement', '');
     ampStoryAdPage./*OK*/innerHTML = `
       <amp-story-grid-layer template="vertical">
-        <h1>First Ad Page</h1>
-        <p>This is the first ad shown in this story.</p>
+        <h1>Ad Page #${id}</h1>
+        <p>This is ad #${id} shown in this story.</p>
       </amp-story-grid-layer>
       `;
     return ampStoryAdPage;
   }
+
+
+  /**
+   * Get array containing all pages of associated story
+   * @private
+   */
+  getPages_() {
+    if (!this.pages_) {
+      this.pages_ = this.ampStory_.getPages();
```

Comment:
I would expect caching this to be dangerous long-term, since your `pages_` array can get stale if we modify the `pages_` in amp-story.

Reply:
good point will change.

Reply:
removed